teps/0099-parameters-in-script.md

```diff
@@ -0,0 +1,572 @@
+
+Users need a way to safely and reliably substitute `Parameters` in `Scripts` in `Steps`
+of their `Tasks`. This would make it easier to use Tekton without learning advanced topics,
+while still using it reliably and safely.
+
+### Requirements
+
+1. **[Secure by default in v1][secure-by-default]**: Out of the box
+   *Tekton Pipelines* v1 should disallow common anti-patterns that
+   can render a default install insecure.
+1. **Easy to use**: Whatever replaces the direct use of `$(params.X)` in
+   `script` should be unsurprising and memorable to ease friction in user
+   migration.
+1. **Automatable**: A script or program should be able to convert `Tasks`
+   to the new approach. This might be impossible for all cases (e.g. if an
+   unrecognized shebang has been used in the script) but a best-effort
+   implementation should be provided.
+1. **Backwards-compatible in v1beta1**. If a user already has `Tasks`
+   that use `$(params.X)` syntax in `script` then we should continue supporting
+   that usage in v1beta1 resources. Ideally, Tekton Pipelines would warn the user
+   about it somehow.
+1. **Provide an escape hatch in v1**. If a cluster operator or admin decides
+   that they want to continue allowing `Parameter` variables to be directly injectable
+   to `script` fields of `Steps` there should be some way for them to configure that.
+
+## Alternatives
+
+### 1. Implicitly Project `Parameters` as Environment Variables in `Steps`
+
+We have projected `Parameters` as environment variables in the `git-clone` Task in the
+Catalog and in our own plumbing repository. We can use this approach, and simplify the
+syntax by implicitly projecting `Parameters` as environment variables in `Steps`.
+The environment variable should be prefixed with `PARAM_` to make it clear and predictable.
+
+```yaml
+  # Before
+  steps:
+  - name: foo
+    image: myimage
+    script: |
+      echo $(params.bar)
+  
+  # After
+  steps:
+  - name: foo
+    image: myimage
+    script: |
+      echo ${PARAM_BAR}
+      
+  # After - Resolved
+  steps:
+  - name: foo
+    image: myimage
+    env:
+    - name: PARAM_BAR
+      value: $(params.bar)
+    script: |
+      echo ${PARAM_BAR}      
+```
+
+##### Discussion
+
+This approach mutates the specification, but it is something we're already doing
+with implicit `Parameters` per [TEP-0023][tep-0023].
+
+### 2. Implicitly Project `Parameters` into the Filesystem of `Steps`
+
+Write `Parameters` in files - `/tekton/parameters/in` - mounted into `Steps`
+in the same way that `Results` are supplied. The file path for the param
+would be made available via a variable (possibly reusing `$(params.<name>)`.
+
+
+```yaml
+  # Before
+  steps:
+  - name: foo
+    image: myimage
+    script: |
+      echo $(params.bar)
+  
+  # After
+  steps:
+  - name: foo
+    image: myimage
+    script: |
+      # note: echo becomes cat because we now print
+      # the contents of a file
+      cat $(params.bar)
```

Comment:
I hear you on the real-world examples and will update the content here to use `git clone` as you suggest. Just want to respond to one thing you mentioned:

> While removing a trailing linebreak characters is often the right thing — sometimes it is not, and knowing when to do what is important.

Our aim with this TEP is not to "fix" bash scripting, or even make it easier really. As you mentioned elsewhere we have to lean a little bit on script authors to be knowledgeable about the language they're using.

We're really looking for something like an "as-close-as-possible-to-existing-behaviour" solution that allows us to stop injecting content into executable contexts by default. I've taken the `read`-based approach you described so that we continue respecting any leading and trailing whitespace, similar to if the content of the param were injected directly.

Your example elsewhere of prepared statements in SQL is spot on for the intentions behind this TEP I think!